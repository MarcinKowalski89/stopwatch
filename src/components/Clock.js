import React from 'react';
import PropTypes from 'prop-types';

const Clock = (props) => {
  const h = props.time.getHours();
  const m = props.time.getMinutes();
  const s = props.time.getSeconds();
  const hdeg = (h * 30) + (m / 2);
  const mdeg = (m * 6) + (s / 10);
  const sdeg = s * 6;

  const hoursDeg = {
    transform: `rotate(${hdeg}deg)`,
  };

  const minutesDeg = {
    transform: `rotate(${mdeg}deg)`,
  };

  const secondsDeg = {
    transform: `rotate(${sdeg}deg)`,
  };

  return (
    <div className="clock-wrapper">
      <div className="clock">
        <div
          className="hours"
          style={hoursDeg}
        />
        <div
          className="minutes"
          style={minutesDeg}
        />
        <div
          className="seconds"
          style={secondsDeg}
        />
      </div>
      <h4 className="city">{props.city}</h4>
    </div>
  );
};

Clock.propTypes = {
  time: PropTypes.instanceOf(Date).isRequired,
  city: PropTypes.string.isRequired,
};

export default Clock;
