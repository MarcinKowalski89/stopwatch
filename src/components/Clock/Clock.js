import React from 'react';
import PropTypes from 'prop-types';

import timeHelper from '../../../helpers/time';

class Clock extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      time: new Date(new Date()
        .toLocaleString('en-US', { timeZone: this.props.zoneName })),
    };
  }

  componentDidMount() {
    this.clock = setInterval(() => this.tick(), 100);
  }

  componentWillUnmount() {
    clearInterval(this.clock);
  }

  getCityName() {
    return this.props.zoneName.split('/').splice(-1);
  }

  displayProperCityName() {
    return this.getCityName().toString().replace(/_|-/g, ' ');
  }

  tick() {
    this.setState({
      time: new Date(new Date()
        .toLocaleString('en-US', { timeZone: this.props.zoneName })),
    });
  }

  render() {
    const { h, m, s } = timeHelper(this.state.time);
    const hdeg = (h * 30) + (m / 2);
    const mdeg = (m * 6) + (s / 10);
    const sdeg = s * 6;

    const hoursDeg = {
      transform: `rotate(${hdeg}deg)`,
    };

    const minutesDeg = {
      transform: `rotate(${mdeg}deg)`,
    };

    const secondsDeg = {
      transform: `rotate(${sdeg}deg)`,
    };
    return (
      <div className="clock-wrapper">
        <div className="clock">
          <button
            className="btn btn-danger"
            onClick={() => this.props.hideClock(this.getCityName())}
          >
            X
          </button>
          <div
            className="hours"
            style={hoursDeg}
          />
          <div
            className="minutes"
            style={minutesDeg}
          />
          <div
            className="seconds"
            style={secondsDeg}
          />
        </div>
        <h4 className="city">{this.displayProperCityName()}</h4>
      </div>
    );
  }
}

Clock.propTypes = {
  hideClock: PropTypes.func.isRequired,
  zoneName: PropTypes.string.isRequired,
};

export default Clock;
