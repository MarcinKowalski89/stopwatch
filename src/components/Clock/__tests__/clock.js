import React from 'react';
import { shallow } from 'enzyme';

import Clock from '../Clock';

test('Clock should match default snapshot', () => {
  const constantDate = new Date('2017-06-13T04:41:20')

  Date = class extends Date {
    constructor() {
      super();
      return constantDate;
    }
  };

  const component = shallow(<Clock
    hideClock={() => null}
    city="Warsaw"
    locale="en-US"
    timeZone="Europe/Warsaw"
  />);

  expect(component).toMatchSnapshot();
});
