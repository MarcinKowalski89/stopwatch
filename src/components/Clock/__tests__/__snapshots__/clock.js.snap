// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clock should match default snapshot 1`] = `
<div
  className="clock-wrapper"
>
  <div
    className="clock"
  >
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      X
    </button>
    <div
      className="hours"
      style={
        Object {
          "transform": "rotate(200.5deg)",
        }
      }
    />
    <div
      className="minutes"
      style={
        Object {
          "transform": "rotate(248deg)",
        }
      }
    />
    <div
      className="seconds"
      style={
        Object {
          "transform": "rotate(120deg)",
        }
      }
    />
  </div>
  <h4
    className="city"
  >
    Warsaw
  </h4>
</div>
`;
